### ISSUE TITLE
Twisted should move to a src/ directory layout

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8214]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8214|
|Type|enhancement|
|Created|2016-02-19 11:52:09Z|
|Branch|https://github.com/twisted/twisted/tree/src-move-8214-5|

This is current best practice (https://hynek.me/articles/testing-packaging/) and fixes quite a few issues:

- tox having to cd to a different directory to run the tests without running against the checkout accidentally (all of the ```toxrundir``` lines)
- reduce confusion as scripts will put stuff in the local checkout dir, instead of sometimes random dirs in ```build/``` (like ```.coverage``` files, or apidocs)
- allow us to have different packages in the repo that work in lockstep (see [#7945](https://github.com/twisted/twisted/issues/7945))

<details><summary>Searchable metadata</summary>

```
trac-id__8214 8214
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__None None
branch__src_move_8214_5 src-move-8214-5
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1455882729084077 1455882729084077
changetime__1471480431447809 1471480431447809
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [46887]) Branching to src-move-8214.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

This depends on [#7985](https://github.com/twisted/twisted/issues/7985) to make the py3 tests pass on an installed Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [46986]) Branching to src-move-8214-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [47107]) Branching to src-move-8214-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

Requirement for the buildbots to work with this patch: https://github.com/twisted-infra/braid/pull/191

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

Will this retain the svn commit history? Isn't it better to svn mv the code, it now looks like you're doing a copy.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

It should -- but I'm using the git-svn bridge :(

I will have to try it using svn mv.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [47113]) Branching to src-move-8214-4.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Nope, svn mv still breaks git-svn. :(

Looks like this'll have to wait until the git move.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set owner to @hawkowl|
|-|-|

We're on Git now. Time to revisit this.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

I am fairly happy with this, please review. Builders are spinning.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

See PR: https://github.com/twisted/twisted/pull/401

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

Looks good, but what is the point of this move?

Why are you only `fairly happy` with this branch. Why is not right with this branch and stops you for being fully happy with it :)

It  would help to understand what are the drawbacks that you have identified or the reasons why you are not very happy with the branch :)

------------

The link to tox.ini line 13 gave me no clue about why is this change is needed :) 

The bin/trial script is still importing the Twisted from the 'src/' folder. How is this helping with running tests against an installed Twisted version?

Leaving it for review so that another review can check it.

I don't see any harm in merging this branch, but in the same time I don't see the benefits.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl commented|
|-|-|

I'm 'fairly happy' in that the builds are green, but I am having to move every file in the repo :P

Line 13 is now https://github.com/twisted/twisted/blob/trunk/tox.ini#L16, the changedir hacks. It also helps with an installed version because the 'trial' (not 'bin/trial'!) you run in a virtualenv is now the installed twisted, not the random one on the path.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_17"></a>|@adiroiban commented|
|-|-|

Hm... but the changedir hack is still in the tox.ini file in the proposed branch... and I can see that bin/trial is still used ... but maybe this will be changes in a follow up ticket... but this is not clear to me and not follow up ticket is referenced here

Somehow I don't get it... but don't worry :) leaving this in the review queue so that another dev can take a look

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl commented|
|-|-|

Right, those aren't needed, so I've removed them now. They were overriding the top-level one.

As far as using bin/trial, I've tried to remove that as well.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

The github 'update branch' button didn't work, but I just merged this (there were no actual conflicts according to git) and pushed to it (236 commits behind!)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

I added a little throwaway script to the branch which automates the process of doing the move of files from `twisted/` to `src/twisted/`; this should help both with keeping it up to date with trunk and also helping deal with the post-merge fallout.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks hawkowl!

  1. As Adi pointed out, I think I know the motivation for doing this, but we should update the description of this ticket to clearly explain what the point is.  Especially if you have a list of concrete problems that are caused by this. For example, I'm not sure I totally get why it will make removing that tox line easier...
  2. Things related to coverage XML generation seem to be failing.  I don't know why those fail but other things pass, the errors I've seen are empty.

Hopefully the script I added will make it easier to keep it up to date as you poke at those.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_23"></a>|@hawkowl removed owner|
|-|-|

I have managed to break the merge checker :D

I've updated the branch to match trunk, and also fixed up the description. Buildbot is green.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_24"></a>|@hawkowl commented|
|-|-|

To check the diff: ```git fetch && git diff origin/trunk..origin/src-move-8214-5 --diff-filter=M```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_25"></a>|@adiroiban set owner to @hawkowl|
|-|-|

All good. Thanks!

Let'd do it :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_26"></a>|@hawkowl set status to `closed`|
|-|-|

In changeset ed075dad7339456c47fe43a51a12ee4b5910e0fd
```
#!CommitTicketReference repository="" revision="ed075dad7339456c47fe43a51a12ee4b5910e0fd"
Merge src-move-8214-5: Move twisted under a src/ directory

Author: hawkowl, glyph
Reviewers: glyph, adiroiban
Fixes: #8214
```
